### PR TITLE
Bugfixes + tests for tasks with rules

### PR DESCRIPTION
**Pull request by Buckram123**
- Fixed bug of `GetAgentTasks` returning wrong amount of tasks (It was returning next block of task, instead of amount of tasks for this user) TODO: need more tests with this
- Moved `Rule` struct to it's own lib
- cw-croncat-core no longer creating wasm
- Few more tests for tasks with rules
- Few renames in cw_rules for cleaner snake_case

### CODE REVIEW

**Comment by Buckram123** [comment]
Found a few rules-related bugs, fixing it and creating tests to cover those

File: contracts/cw-croncat/src/helpers.rs [hunk 1]
```diff
@@ -228,7 +228,11 @@ impl<'a> CwCroncat<'a> {
                 config.available_balance.cw20.find_checked_sub(&sent)?;
             };
             self.config.save(storage, &config)?;
-            self.tasks.save(storage, &task_hash, &task)?;
+            if task.with_rules() {
```
Comment (mikedotexe) [inline contracts/cw-croncat/src/helpers.rs:231]: ❤️

File: contracts/cw-croncat/src/slots.rs [hunk 2]
```diff
@@ -63,8 +63,8 @@ impl<'a> CwCroncat<'a> {
         kind: &SlotType,
     ) -> Option<Vec<u8>> {
         let store = match kind {
-            SlotType::Block => self.block_slots.clone(),
-            SlotType::Cron => self.time_slots.clone(),
+            SlotType::Block => &self.block_slots,
```
Comment (mikedotexe) [inline contracts/cw-croncat/src/slots.rs:66]: good eye 🚀

File: contracts/cw-croncat/src/slots.rs [hunk 3]
```diff
@@ -74,24 +74,41 @@ impl<'a> CwCroncat<'a> {
...
-    /// Used in cases where there are empty slots or failed txns
-    fn clean_slot(&mut self, storage: &mut dyn Storage, slot: &u64, kind: &SlotType) {
-        let store = match kind {
-            SlotType::Block => self.block_slots.clone(),
-            SlotType::Cron => self.time_slots.clone(),
-        };
-
-        store.remove(storage, *slot);
-    }
+    //     /// Gets 1 slot hash item, and removes the hash from storage
```
Comment (mikedotexe) [inline contracts/cw-croncat/src/slots.rs:85]: Are we thinking we'll do this in another ticket?

File: contracts/cw-croncat/src/state.rs [hunk 4]
```diff
@@ -112,12 +112,12 @@ pub struct CwCroncat<'a> {
     /// this is done instead of forcing a block height into a range of timestamps for reliability
     pub block_slots: Map<'a, u64, Vec<Vec<u8>>>,
 
-    pub tasks_with_rules: IndexedMap<'a, Vec<u8>, Task, TaskIndexes<'a>>,
+    pub tasks_with_rules: IndexedMap<'a, &'a [u8], Task, TaskIndexes<'a>>,
```
Comment (mikedotexe) [inline contracts/cw-croncat/src/state.rs:115]: I was just reading through the Tokio documentation here:
Screen Shot 2022-09-20 at 6 31 29 PM

https://tokio.rs/tokio/tutorial/shared-state#add-bytes-dependency

I'm not suggesting we change it, but thought you may find that insightful given the change to a byte array I'm seeing here.))

File: contracts/cw-croncat/src/tasks.rs [hunk 5]
```diff
@@ -1395,6 +1403,109 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn check_task_with_rules_and_without_create_success() -> StdResult<()> {
```
Comment (mikedotexe) [inline contracts/cw-croncat/src/tasks.rs:1407]: Thank you for adding this test!

File: contracts/cw-rules/src/contract.rs [hunk 6]
```diff
@@ -58,15 +58,15 @@ pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
...
+        QueryMsg::GetCw20Balance {
             cw20_contract,
             address,
         } => to_binary(&query_get_cw20_balance(deps, cw20_contract, address)?),
         QueryMsg::HasBalanceGte(HasBalanceGte {
             address,
             required_balance,
         }) => to_binary(&query_has_balance_gte(deps, address, required_balance)?),
-        QueryMsg::CheckOwnerOfNFT(CheckOwnerOfNFT {
+        QueryMsg::CheckOwnerOfNft(CheckOwnerOfNft {
```
Comment (mikedotexe) [inline contracts/cw-rules/src/contract.rs:69]: Haha, snake case is awful otherwise, huh 😆

**Review by mikedotexe — APPROVED** [review]

**Pull request merged**